src/particles/diagnostics/BeamRelevant.cpp

```diff
@@ -0,0 +1,225 @@
+#include <AMReX_BLProfiler.H>           // for TinyProfiler
+
+using namespace amrex;
+
+namespace impactx
+{
+
+    namespace diagnostics{
+
+
+        void compute_beam_relevant(ImpactXParticleContainer const & pc, int const step) {
+
+            BL_PROFILE("impactX::diagnostics::compute_beam_relevant");
+
+            // physical constants and reference quantities
+            amrex::ParticleReal constexpr c0_SI = 2.99792458e8;  // TODO move out
+            // inverse of speed of light squared
+            amrex::Real constexpr inv_c2 = 1.0_rt / (c0_SI * c0_SI);
+
+            // preparing to access reference particle data: RefPart
+            RefPart const ref_part = pc.GetRefParticle();
+
+            ParticleReal const m_MeV = ref_part.mass_MeV();
+            ParticleReal const q_qe = ref_part.charge_qe();
+
+            // preparing access to particle data: AoS and SoA
+            using PTDType = typename ImpactXParticleContainer::ParticleTileType::ConstParticleTileDataType;
+
+            amrex::ReduceOps<ReduceOpSum,ReduceOpSum,ReduceOpSum,ReduceOpSum,ReduceOpSum,
+            ReduceOpSum,ReduceOpSum,ReduceOpSum> reduce_ops;
+
+            auto r = amrex::ParticleReduce<amrex::ReduceData<ParticleReal,ParticleReal,
+            ParticleReal,ParticleReal,ParticleReal,ParticleReal,ParticleReal,ParticleReal>>(
+                    pc,
+                    [=] AMREX_GPU_DEVICE(const PTDType& ptd, const int i) noexcept -> amrex::GpuTuple
+                            <ParticleReal,ParticleReal,ParticleReal,ParticleReal,ParticleReal,
+                            ParticleReal,ParticleReal,ParticleReal>
+                    {
+                        // access AoS particle position data
+                        const ParticleReal p_pos0 = ptd.m_aos[i].pos(0);
+                        const ParticleReal p_pos1 = ptd.m_aos[i].pos(1);
+                        const ParticleReal p_pos2 = ptd.m_aos[i].pos(2);
+
+                        // access SoA particle momentum data and weighting
+                        const ParticleReal p_w = ptd.m_rdata[RealSoA::w][i];
+                        const ParticleReal p_ux = ptd.m_rdata[RealSoA::ux][i];
+                        const ParticleReal p_uy = ptd.m_rdata[RealSoA::uy][i];
+                        const ParticleReal p_ut = ptd.m_rdata[RealSoA::pt][i];
+
+                        const ParticleReal p_us = p_ux*p_ux + p_uy*p_uy + p_ut*p_ut;
+                        // prepare mean position values
+                        const ParticleReal p_x_mean = p_pos0*p_w;
+                        const ParticleReal p_y_mean = p_pos1*p_w;
+                        const ParticleReal p_t_mean = p_pos2*p_w;
+
+                        const ParticleReal p_ux_mean = p_ux*p_w;
+                        const ParticleReal p_uy_mean = p_uy*p_w;
+                        const ParticleReal p_ut_mean = p_ut*p_w;
+                        const ParticleReal p_gm_mean = std::sqrt(1.0_rt+p_us*inv_c2)*p_w;
```

Comment:
I'm not sure about this line.  What is p_gm intended to represent?  Internally, the momenta should be dimensionless, so I don't think 1 + p_us*inv_c2 can work for dimensional reasons.

Reply:
Thanks! I removed those :)